### ISSUE TITLE
Any plan to add an Embedding operator?

### ISSUE THREAD

**Issue by yaoyuannnn** [onnx/onnx#3685]
I wanted to ask if there's any plan to add an ONNX Embedding operator?

**Comment by linkerzhang** [comment]
Good idea. This may make ONNX to be usable in many recommendation models.

**Comment by yaoyuannnn** [comment]
Yes. Otherwise the frameworks will need to do some sort of pattern matching to find the low-level ONNX operators and fuse them. Let me know if this operator is under the radar, we can then plan accordingly.

**Comment by linkerzhang** [comment]
+ @ebarsoum who's running operator sig.

@yaoyuannnn you may also want to send a PR to propose such an operator or function please.

**Comment by askhade** [comment]
Right now this op is not under consideration. As @linkerzhang suggested please send a PR for your proposal. 
@postrational and @gramalingam are OP SIG leads. Once you propose a PR please tag it with operator label.

**Comment by gramalingam** [comment]
Can you clarify what the embedding operator would do? E.g., vs. using matrix-multiplication (possibly with casts from ints to floats)?

**Comment by yaoyuannnn** [comment]
Thanks for the replies. I was thinking if ONNX could have an [EmbeddingBag](https://pytorch.org/docs/stable/generated/torch.nn.functional.embedding_bag.html?highlight=embedding#torch.nn.functional.embedding_bag) operator like PyTorch, which in addition to doing embedding lookups, it also supports a mode to compute the sum or mean of the embeddings. A standalone [Embedding](https://pytorch.org/docs/stable/generated/torch.nn.functional.embedding.html?highlight=embedding#torch.nn.functional.embeddingg) operator should also be fine. @gramalingam I don't think the ONNX spec would suggest how an operator should be implemented (like using matrix-multiply or not), right? The weights can be in any data type, so no casts should be needed.

**Comment by garymm** [comment]
What's the detailed use case here? Are you using a converter?

`torch.onnx.export` supports embedding:

https://github.com/pytorch/pytorch/blob/ef5e2481ce353255d5b23d6930426c6486aa8721/torch/onnx/symbolic_opset9.py#L482


And embedding_bag:
https://github.com/pytorch/pytorch/blob/ef5e2481ce353255d5b23d6930426c6486aa8721/torch/onnx/symbolic_opset10.py#L242

**Comment by yaoyuannnn** [comment]
> What's the detailed use case here? Are you using a converter?
> 
> `torch.onnx.export` supports embedding:
> 
> https://github.com/pytorch/pytorch/blob/ef5e2481ce353255d5b23d6930426c6486aa8721/torch/onnx/symbolic_opset9.py#L482
> 
> And embedding_bag: https://github.com/pytorch/pytorch/blob/ef5e2481ce353255d5b23d6930426c6486aa8721/torch/onnx/symbolic_opset10.py#L242

We would like to see an ONNX embedding operator instead of letting the framework break up an EmbeddingBag into low-level ONNX ops. One use case would be TensorRT, which takes ONNX as input, and it would have to pattern match these low-level ops and fuse them back into a single native operator.

**Comment by askhade** [comment]
In that case we can add a function op (which is composed of primitive ops) this way any runtime which has a specialized kernel can simply dispatch it and others can inline the primitive ops. Are you willing to contribute?

**Comment by yaoyuannnn** [comment]
Thanks. Can you link me to any example of a function op?

**Comment by garymm** [comment]
https://github.com/onnx/onnx/pull/3431

**Comment by yaoyuannnn** [comment]
Thanks! If I understand it correctly, a function op will be a `FunctionProto` and the user can get the function body that contains the low-level ops.

**Comment by askhade** [comment]
> Thanks! If I understand it correctly, a function op will be a `FunctionProto` and the user can get the function body that contains the low-level ops.

Right. The Op is onnx schema defined op with an additional functionproto - https://github.com/onnx/onnx/blob/master/onnx/defs/generator/defs.cc#L825

**Comment by askhade** [comment]
I will assign this to you so that we can track progress for this.

**Comment by linkerzhang** [comment]
@yaoyuannnn any progress on contributing an embedding function please?

**Comment by yaoyuannnn** [comment]
@linkerzhang Please find a better person for this task as we don't currently have any plan to contribute an embedding function op.
